File: ProyectoFinal/ProyectoFinal/AlgoritmosdeOdenamiento/BurbujaBidireccional.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using System.Diagnostics;

namespace ProyectoFinal.AlgoritmosdeOdenamiento
{
    public partial class BurbujaBidireccional : Form
    {
        public BurbujaBidireccional()
        {
            InitializeComponent();
        }

        int[] vector;
        int n;
        int min;
        int max;
        int i;
        Stopwatch contador = new Stopwatch();

        public void GenerarDatos(int n, int min, int max)
        {
            vector = new int[n];
            Random aleatorio = new Random();
            for (i = 0; i < vector.Length; i++)
            {
                vector[i] = aleatorio.Next(min, max);

            }
        }

        public void Metodo(int n)
        {

            int der = vector.Length - 1;
            int izq = 0;
            int ultimo = 0;
            int aux;
            int comparaciones = 0;
            int intercambios = 0;
            do
            {
                //recorrido a la derecha
                for (int i = izq; i < der; i++)
                {
                    comparaciones++;
                    if (vector[i] > vector[i + 1])
                    {
                        aux = vector[i];
                        vector[i] = vector[i + 1];
                        vector[i + 1] = aux;
                        ultimo = i;
                        intercambios++;
                    }
                }
                der = ultimo;

                //recorrido a la izq
                for (int j = der; j > izq; j--)
                {
                    comparaciones++;
                    if (vector[j - 1] > vector[j])
                    {
                        aux = vector[j];
                        vector[j] = vector[j - 1];
                        vector[j - 1] = aux;
                        ultimo = j;
                        intercambios++;
                    }

                }
                izq = ultimo;

            } while (izq < der);

            lblComparaciones.Text = comparaciones.ToString() + " Comparaciones";
            lblIntercambios.Text = intercambios.ToString() + " Intercambios";
            return;
        }

        public void Mostrar(ListBox l)
        {
            for (i = 0; i < vector.Length; i++)
            {
                l.Items.Add(vector[i]);
            }
        }

        private void btnGenerar_Click(object sender, EventArgs e)
        {
            try
            {
                lbNum.Items.Clear();
                lbOrd.Items.Clear();
                n = int.Parse(txtNum.Text);
                min = int.Parse(txtMin.Text);
                max = int.Parse(txtMax.Text);
                GenerarDatos(n, min, max);
                Mostrar(lbNum);
                btnGenerar.Enabled = false;
                btnOrdenar.Enabled = true;
            }
            catch
            {
                MessageBox.Show("Introduzca un numero valido");
            }
        }

        private void btnOrdenar_Click(object sender, EventArgs e)
        {
            contador.Restart();
            Metodo(n);
            contador.Stop();
            lblOrdenar.Text = contador.Elapsed.TotalMilliseconds.ToString() + " Milisegundos";
            Mostrar(lbOrd);
            btnGenerar.Enabled = true;
            btnOrdenar.Enabled = false;
        }

        private void algoritmosToolStripMenuItem_Click(object sender, EventArgs e)
        {
            MenuAlgoritmos paransaek = new MenuAlgoritmos();
            paransaek.Show();
            this.Close();
        }

        private void estructurasToolStripMenuItem_Click(object sender, EventArgs e)
        {
            MenuEstructuras ppalgansaek = new MenuEstructuras();
            ppalgansaek.Show();
            this.Close();
        }

        private void salirToolStripMenuItem_Click(object sender, EventArgs e)
        {
            Application.Exit();
        }
    }
}

File: ProyectoFinalVB/ProyectoFinalVB/AlgoritmosdeOrdenamiento/BurbujaBidireccional.vb
Imports System.Diagnostics
Public Class BurbujaBidireccional
    Dim vector As Integer()
    Dim n As Integer
    Dim min As Integer
    Dim max As Integer
    Dim i As Integer
    Dim contador As Stopwatch = New Stopwatch()

    Public Sub GenerarDatos(ByVal n As Integer, ByVal min As Integer, ByVal max As Integer)
        vector = New Integer(n - 1) {}
        Dim aleatorio As Random = New Random()
        For i = 0 To vector.Length - 1
            vector(i) = aleatorio.[Next](min, max)
        Next
    End Sub

    'arreglar error en el metodo
    Public Sub Metodo(ByVal n)
        Dim der As Integer = vector.Length - 1
        Dim izq As Integer = 0
        Dim ultimo As Integer = 0
        Dim aux As Integer
        Dim comparaciones As Integer = 0
        Dim intercambios As Integer = 0

        Do
            For i As Integer = izq To der - 1
                comparaciones += 1
                If vector(i) > vector(i + 1) Then
                    aux = vector(i)
                    vector(i) = vector(i + 1)
                    vector(i + 1) = aux
                    ultimo = i
                    intercambios += 1
                End If
            Next
            der = ultimo

            For j As Integer = der To izq + 1
                comparaciones += 1
                If vector(j - 1) > vector(j) Then
                    aux = vector(j)
                    vector(j) = vector(j - 1)
                    vector(j - 1) = aux
                    ultimo = j
                    intercambios += 1
                End If
            Next
            izq = ultimo

        Loop While izq < der

        lblComparaciones.Text = comparaciones.ToString() & " Comparaciones"
        lblIntercambios.Text = intercambios.ToString() & " Intercambios"
        Return
    End Sub

    Public Sub Mostrar(ByVal l As ListBox)
        For i = 0 To vector.Length - 1
            l.Items.Add(vector(i))
        Next
    End Sub

    Private Sub btnGenerar_Click(sender As Object, e As EventArgs) Handles btnGenerar.Click
        Try
            lbNum.Items.Clear()
            lbOrd.Items.Clear()
            n = Integer.Parse(txtNum.Text)
            min = Integer.Parse(txtMin.Text)
            max = Integer.Parse(txtMax.Text)
            GenerarDatos(n, min, max)
            Mostrar(lbNum)
            btnGenerar.Enabled = False
            btnOrdenar.Enabled = True
        Catch ex As Exception
            MessageBox.Show("Introduzca un numero valido")
        End Try
    End Sub

    Private Sub btnOrdenar_Click(sender As Object, e As EventArgs) Handles btnOrdenar.Click
        contador.Restart()
        Metodo(n)
        contador.Stop()
        lblOrdenar.Text = contador.Elapsed.TotalMilliseconds.ToString() & " Milisegundos"
        Mostrar(lbOrd)
        btnGenerar.Enabled = True
        btnOrdenar.Enabled = False
    End Sub

    Private Sub AlgoritmosToolStripMenuItem_Click(sender As Object, e As EventArgs) Handles AlgoritmosToolStripMenuItem.Click
        Dim paransaek As MenuAlgoritmos = New MenuAlgoritmos()
        paransaek.Show()
        Me.Close()
    End Sub

    Private Sub EstructurasToolStripMenuItem_Click(sender As Object, e As EventArgs) Handles EstructurasToolStripMenuItem.Click
        Dim ppalgansaek As MenuEstructuras = New MenuEstructuras()
        ppalgansaek.Show()
        Me.Close()
    End Sub

    Private Sub SalirToolStripMenuItem_Click(sender As Object, e As EventArgs) Handles SalirToolStripMenuItem.Click
        Application.Exit()
    End Sub
End Class